activation/post_states.go

```diff
@@ -0,0 +1,103 @@
+package activation
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	pb "github.com/spacemeshos/api/release/go/spacemesh/v1"
+	"go.uber.org/zap"
+
+	"github.com/spacemeshos/go-spacemesh/common/types"
+	"github.com/spacemeshos/go-spacemesh/events"
+)
+
+type postStates struct {
+	log            *zap.Logger
+	mu             sync.RWMutex
+	states         map[types.NodeID]postState
+	watchingStates sync.Once
+}
+
+func newPostStates(log *zap.Logger) postStates {
+	return postStates{
+		log:    log,
+		states: make(map[types.NodeID]postState),
+	}
+}
+
+type postState int
+
+// Values must match the ones described in github.com/spacemeshos/go-spacemesh/api/grpcserver/interface.go
+// under the postState interface.
+const (
+	stateIdle postState = iota
+	stateProving
+)
+
+func (s postState) String() string {
+	switch s {
+	case stateIdle:
+		return "idle"
+	case stateProving:
+		return "proving"
+	default:
+		panic(fmt.Sprintf("unknown post state %d", s))
+	}
+}
+
+func (s *postStates) set(id types.NodeID, state postState) {
+	s.mu.Lock()
+	s.states[id] = state
+	s.mu.Unlock()
+
+	s.log.Info("post state changed", zap.Stringer("id", id), zap.Stringer("state", state))
+}
+
+func (s *postStates) get() map[types.NodeID]int {
+	s.mu.RLock()
+	defer s.mu.RUnlock()
+	states := make(map[types.NodeID]int, len(s.states))
+	for id, state := range s.states {
+		states[id] = int(state)
+	}
+	return states
+}
+
+func (s *postStates) watchEvents(ctx context.Context) error {
```

Comment:
I think this method should be synchronous and the caller should call it in a go routine with `errgroup.Go`. Reason being that a function that takes a `context.Context` should not depend on the context any more after it returns:

```go
ctx, cancel := context.WithCancel(context.Background())
defer cancel()
s.watchEvents(ctx)
```

This starts and immediately stops the event watching but should be a legitimate way to call the function (even if that would mean it blocks forever).